File: packages/cli/src/proxy/handler.ts
// src/proxy/handler.ts
import type { ClaudeCodeProfile } from "../types.js";
import { SWIXTER_PROXY_AUTH_TOKEN } from "../constants/proxy.js";
// Load transformer registrations (side-effect: populates TRANSFORMER_REGISTRY)
import "./transform/streaming/openai-chat-to-anthropic.js";
import {
  getGeneralProxyModel,
  isSwixterClaudeProxyMarker,
  resolveSwixterClaudeProxyMarker,
} from "../utils/model-helper.js";
import { ProxyRouter } from "./router.js";
import { ProxyForwarder } from "./forwarder.js";
import { CircuitBreaker } from "./circuit-breaker.js";
import { getGroup, getActiveGroup } from "../groups/manager.js";
import { getProfile } from "../config/manager.js";
import { createProxyLogger, type ProxyLogger } from "./logger.js";
import { getPresetById } from "../providers/presets.js";
import {
  inferClientFormat,
  inferTargetApiFormat,
  getTransformer,
  transformRequest,
  transformResponse,
  transformStream,
} from "./transform/index.js";
import type { TransformContext } from "./transform/types.js";

export class ProxyHandler {
  private router: ProxyRouter;
  private forwarder: ProxyForwarder;
  private circuitBreaker: CircuitBreaker;
  private timeoutMs?: number;
  private instanceId: string;
  private groupName?: string;
  private profileName?: string;
  private logger: ProxyLogger;

  constructor(timeoutMs?: number, instanceId?: string, groupName?: string, profileName?: string) {
    this.router = new ProxyRouter();
    this.forwarder = new ProxyForwarder();
    this.circuitBreaker = new CircuitBreaker();
    this.timeoutMs = timeoutMs;
    this.instanceId = instanceId || "default";
    this.groupName = groupName;
    this.profileName = profileName;
    this.logger = createProxyLogger(this.instanceId);
    this.setupRoutes();
  }

  private setupRoutes(): void {
    // OpenAI Chat format
    this.router.register("/v1/chat/completions", this.handleChatCompletions.bind(this));

    // Anthropic Messages format
    this.router.register("/v1/messages", this.handleMessages.bind(this));

    // Anthropic Responses format
    this.router.register("/v1/responses", this.handleResponses.bind(this));

    // Anthropic compatible
    this.router.register("/anthropic/*", this.handleAnthropic.bind(this));

    // Health check
    this.router.register("/health", this.handleHealth.bind(this));
  }

  async handleRequest(request: Request): Promise<Response> {
    const start = Date.now();
    this.logger.info("Incoming request", { method: request.method, path: new URL(request.url).pathname });

    try {
      const authResponse = this.authenticateProxyRequest(request);
      if (authResponse) {
        this.logger.request(request.method, new URL(request.url).pathname, authResponse.status, Date.now() - start);
        return authResponse;
      }

      const response = await this.router.handle(request);
      this.logger.request(request.method, new URL(request.url).pathname, response.status, Date.now() - start);
      return response;
    } catch (error) {
      this.logger.error("Request handling failed", error as Error);
      this.logger.request(request.method, new URL(request.url).pathname, 500, Date.now() - start);
      return new Response("Internal Server Error", { status: 500 });
    }
  }

  private isHealthRequest(request: Request): boolean {
    return new URL(request.url).pathname === "/health";
  }

  private getBearerToken(request: Request): string | null {
    const authHeader = request.headers.get("authorization");
    if (!authHeader?.startsWith("Bearer ")) {
      return null;
    }

    return authHeader.slice("Bearer ".length);
  }

  private authenticateProxyRequest(request: Request): Response | null {
    if (this.isHealthRequest(request)) {
      return null;
    }

    const token = this.getBearerToken(request);
    if (token === SWIXTER_PROXY_AUTH_TOKEN) {
      return null;
    }

    return new Response(JSON.stringify({ error: "Invalid or missing proxy authentication" }), {
      status: 401,
      headers: { "Content-Type": "application/json" },
    });
  }

  private async handleChatCompletions(request: Request): Promise<Response> {
    return this.forwardToProvider(request, "chat");
  }

  private async handleMessages(request: Request): Promise<Response> {
    return this.forwardToProvider(request, "anthropic");
  }

  private async handleResponses(request: Request): Promise<Response> {
    return this.forwardToProvider(request, "anthropic");
  }

  private async handleAnthropic(request: Request): Promise<Response> {
    return this.forwardToProvider(request, "anthropic");
  }

  private handleHealth(): Response {
    return new Response(JSON.stringify({
      status: "ok",
      instanceId: this.instanceId,
      groupName: this.groupName || null,
      timestamp: new Date().toISOString(),
      uptime: process.uptime(),
    }), {
      headers: { "Content-Type": "application/json" },
    });
  }

  private rewriteRequestBodyForProfile(bodyBuffer: ArrayBuffer, profile: ClaudeCodeProfile): Uint8Array {
    try {
      const bodyText = Buffer.from(bodyBuffer).toString("utf-8");
      const parsed = JSON.parse(bodyText);

      if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
        return new Uint8Array(bodyBuffer);
      }

      const currentModel = typeof parsed.model === "string" ? parsed.model : undefined;

      if (isSwixterClaudeProxyMarker(currentModel)) {
        const resolvedModel = resolveSwixterClaudeProxyMarker(currentModel, profile);
        if (resolvedModel) {
          parsed.model = resolvedModel;
          return Buffer.from(JSON.stringify(parsed));
        }
        return new Uint8Array(bodyBuffer);
      }

      const generalModel = getGeneralProxyModel(profile);
      if (generalModel) {
        parsed.model = generalModel;
        return Buffer.from(JSON.stringify(parsed));
      }

      return new Uint8Array(bodyBuffer);
    } catch (error) {
      if (error instanceof Error && error.message.includes("cannot resolve requested proxy model marker")) {
        throw error;
      }
      return new Uint8Array(bodyBuffer);
    }
  }

  private async forwardSingleProfile(request: Request, profileName: string, _format: "chat" | "anthropic"): Promise<Response> {
    const profile = await getProfile(profileName);
    if (!profile) {
      this.logger.warn("Profile not found", { profileName });
      return new Response(JSON.stringify({ error: `Profile not found: ${profileName}` }), {
        status: 503,
        headers: { "Content-Type": "application/json" },
      });
    }

    // Read body once as ArrayBuffer
    let bodyBuffer: ArrayBuffer;
    try {
      bodyBuffer = await request.arrayBuffer();
    } catch (error) {
      this.logger.error("Failed to read request body", error as Error);
      return new Response(JSON.stringify({ error: "Failed to read request body" }), {
        status: 400,
        headers: { "Content-Type": "application/json" },
      });
    }

    const preset = getPresetById(profile.providerId);
    const endpoint = new URL(request.url).pathname + new URL(request.url).search;
    const clientFormat = inferClientFormat(endpoint);
    const targetFormat = inferTargetApiFormat(profile, preset || {} as NonNullable<typeof preset>);

    let transformedBodyBuffer = bodyBuffer;
    let targetEndpoint = endpoint;
    let ctx: TransformContext | null = null;

    if (clientFormat !== targetFormat) {
      try {
        const bodyText = Buffer.from(bodyBuffer).toString("utf-8");
        const parsedBody = bodyText ? JSON.parse(bodyText) : {};

        ctx = {
          endpoint,
          clientFormat,
          targetFormat,
          profile,
          preset: preset || {} as NonNullable<typeof preset>,
          stream: parsedBody.stream === true,
        };

        const transformed = transformRequest(parsedBody, ctx);
        transformedBodyBuffer = Buffer.from(JSON.stringify(transformed.body));
        targetEndpoint = transformed.targetEndpoint || endpoint;
      } catch (error) {
        this.logger.error("Request transform failed, falling back to passthrough", error as Error);
        // Fall back to passthrough — keep original body and endpoint
      }
    }

    const rewrittenBody = this.rewriteRequestBodyForProfile(transformedBodyBuffer, profile);
    const clonedRequest = {
      method: request.method,
      path: targetEndpoint,
      headers: Object.fromEntries(request.headers.entries()),
      body: rewrittenBody,
    };

    try {
      const forwardResponse = await this.forwarder.forward(clonedRequest, profile, this.timeoutMs);
      const isSuccess = forwardResponse.status >= 200 && forwardResponse.status < 300;

      if (!isSuccess) {
        this.logger.warn("Provider returned upstream status", {
          profileName,
          status: forwardResponse.status,
        });
        return new Response(forwardResponse.body, {
          status: forwardResponse.status,
          headers: forwardResponse.headers,
        });
      }

      // Transform response back to client format
      if (ctx && clientFormat !== targetFormat) {
        if (forwardResponse.isStream && forwardResponse.body) {
          const transformedStream = transformStream(forwardResponse.body as ReadableStream<Uint8Array>, ctx);
          return new Response(transformedStream, {
            status: forwardResponse.status,
            headers: forwardResponse.headers,
          });
        }

        try {
          const responseText = forwardResponse.body as string;
          const responseParsed = responseText ? JSON.parse(responseText) : {};
          const transformedResponse = transformResponse(responseParsed, ctx);
          return new Response(JSON.stringify(transformedResponse), {
            status: forwardResponse.status,
            headers: forwardResponse.headers,
          });
        } catch (error) {
          this.logger.error("Response transform failed, returning raw response", error as Error);
          return new Response(forwardResponse.body, {
            status: forwardResponse.status,
            headers: forwardResponse.headers,
          });
        }
      }

      // No transform needed — return directly
      if (forwardResponse.isStream) {
        return new Response(forwardResponse.body as ReadableStream, {
          status: forwardResponse.status,
          headers: forwardResponse.headers,
        });
      }

      return new Response(forwardResponse.body, {
        status: forwardResponse.status,
        headers: forwardResponse.headers,
      });
    } catch (error) {
      this.logger.error("Provider request failed", error as Error, { profileName });
      return new Response(JSON.stringify({ error: (error as Error).message }), {
        status: 502,
        headers: { "Content-Type": "application/json" },
      });
    }
  }

  private async forwardToProvider(request: Request, format: "chat" | "anthropic"): Promise<Response> {
    if (this.profileName) {
      return this.forwardSingleProfile(request, this.profileName, format);
    }

    const group = this.groupName
      ? await getGroup(this.groupName)
      : await getActiveGroup();

    if (!group || group.profiles.length === 0) {
      this.logger.warn("No active group or profiles");
      return new Response(JSON.stringify({ error: "No active group or profiles" }), {
        status: 503,
        headers: { "Content-Type": "application/json" },
      });
    }

    // Read body once as ArrayBuffer for multiple provider attempts
    let bodyBuffer: ArrayBuffer;
    try {
      bodyBuffer = await request.arrayBuffer();
    } catch (error) {
      this.logger.error("Failed to read request body", error as Error);
      return new Response(JSON.stringify({ error: "Failed to read request body" }), {
        status: 400,
        headers: { "Content-Type": "application/json" },
      });
    }

    const errors: string[] = [];
    let lastFailureResponse: Response | null = null;

    for (const profileId of group.profiles) {
      if (!this.circuitBreaker.isAvailable(profileId)) {
        this.logger.info("Skipping unavailable provider", { profileId });
        continue;
      }

      const profile = await getProfile(profileId);
      if (!profile) {
        this.logger.warn("Profile not found", { profileId });
        continue;
      }

      const preset = getPresetById(profile.providerId);

      const endpoint = new URL(request.url).pathname + new URL(request.url).search;
      const clientFormat = inferClientFormat(endpoint);
      const targetFormat = inferTargetApiFormat(profile, preset || {} as NonNullable<typeof preset>);

      // Skip provider if formats don't match and no transformer is available
      if (clientFormat !== targetFormat) {
        if (!getTransformer(clientFormat, targetFormat)) {
          this.logger.info("Skipping provider: no transformer for format pair", { profileId, clientFormat, targetFormat });
          continue;
        }
      }

      let transformedBodyBuffer = bodyBuffer;
      let targetEndpoint = endpoint;
      let ctx: TransformContext | null = null;

      if (clientFormat !== targetFormat) {
        try {
          const bodyText = Buffer.from(bodyBuffer).toString("utf-8");
          const parsedBody = bodyText ? JSON.parse(bodyText) : {};

          ctx = {
            endpoint,
            clientFormat,
            targetFormat,
            profile,
            preset: preset || {} as NonNullable<typeof preset>,
            stream: parsedBody.stream === true,
          };

          const transformed = transformRequest(parsedBody, ctx);
          transformedBodyBuffer = Buffer.from(JSON.stringify(transformed.body));
          targetEndpoint = transformed.targetEndpoint || endpoint;
        } catch (error) {
          this.logger.error("Request transform failed, falling back to passthrough", error as Error);
          // Fall back to passthrough — keep original body and endpoint
        }
      }

      try {
        const rewrittenBody = this.rewriteRequestBodyForProfile(transformedBodyBuffer, profile);
        const clonedRequest = {
          method: request.method,
          path: targetEndpoint,
          headers: Object.fromEntries(request.headers.entries()),
          body: rewrittenBody,
        };

        const forwardResponse = await this.forwarder.forward(clonedRequest, profile, this.timeoutMs);
        const isSuccess = forwardResponse.status >= 200 && forwardResponse.status < 300;
        const shouldTripCircuit = forwardResponse.status >= 500 || forwardResponse.status === 429;

        if (!isSuccess) {
          if (shouldTripCircuit) {
            this.circuitBreaker.recordFailure(profileId);
          }

          errors.push(`${profileId}: upstream returned ${forwardResponse.status}`);
          this.logger.warn("Provider returned upstream status", {
            profileId,
            status: forwardResponse.status,
            fallback: true,
          });

          lastFailureResponse = new Response(forwardResponse.body, {
            status: forwardResponse.status,
            headers: forwardResponse.headers,
          });
          continue;
        }

        this.circuitBreaker.recordSuccess(profileId);

        // Transform response back to client format if needed
        if (ctx && clientFormat !== targetFormat) {
          if (forwardResponse.isStream && forwardResponse.body) {
            const transformedStream = transformStream(forwardResponse.body as ReadableStream<Uint8Array>, ctx);
            return new Response(transformedStream, {
              status: forwardResponse.status,
              headers: forwardResponse.headers,
            });
          }

          try {
            const responseText = forwardResponse.body as string;
            const responseParsed = responseText ? JSON.parse(responseText) : {};
            const transformedResponse = transformResponse(responseParsed, ctx);
            return new Response(JSON.stringify(transformedResponse), {
              status: forwardResponse.status,
              headers: forwardResponse.headers,
            });
          } catch (error) {
            this.logger.error("Response transform failed, returning raw response", error as Error);
            return new Response(forwardResponse.body, {
              status: forwardResponse.status,
              headers: forwardResponse.headers,
            });
          }
        }

        // Return streaming response directly
        if (forwardResponse.isStream) {
          return new Response(forwardResponse.body as ReadableStream, {
            status: forwardResponse.status,
            headers: forwardResponse.headers,
          });
        }

        return new Response(forwardResponse.body, {
          status: forwardResponse.status,
          headers: forwardResponse.headers,
        });
      } catch (error) {
        this.circuitBreaker.recordFailure(profileId);
        errors.push(`${profileId}: ${(error as Error).message}`);
        this.logger.error("Provider request failed", error as Error, { profileId });
        continue;
      }
    }

    this.logger.error("All providers failed", undefined, { errors });
    if (lastFailureResponse) {
      return lastFailureResponse;
    }
    return new Response(JSON.stringify({ error: "All providers failed", details: errors }), {
      status: 503,
      headers: { "Content-Type": "application/json" },
    });
  }

  getCircuitBreaker(): CircuitBreaker {
    return this.circuitBreaker;
  }
}

File: packages/cli/crates/proxy/src/handler.rs
use bytes::Bytes;
use reqwest::header::HeaderMap;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

use futures::Stream;
use serde_json::{json, Value};
use std::pin::Pin;
use swixter_core::config::ConfigManager;
use swixter_core::types::{Group, Profile};

use crate::breaker::CircuitBreaker;
use crate::forwarder::{ForwardBody, ForwardRequest, Forwarder};
use crate::logger::ProxyLogger;
use crate::transform;
use crate::types::{ProxyServerConfig, ProxyStatus};
use crate::SWIXTER_PROXY_AUTH_TOKEN;

pub enum HandlerBody {
    Full(Bytes),
    Stream(Pin<Box<dyn Stream<Item = Result<Bytes, std::io::Error>> + Send>>),
}

pub struct HandlerResponse {
    pub status: u16,
    pub headers: HeaderMap,
    pub body: HandlerBody,
}

impl HandlerResponse {
    fn json(status: u16, body: Value) -> Self {
        let mut headers = HeaderMap::new();
        headers.insert("content-type", "application/json".parse().unwrap());
        Self {
            status,
            headers,
            body: HandlerBody::Full(Bytes::from(serde_json::to_vec(&body).unwrap())),
        }
    }

    fn text(status: u16, body: &'static str) -> Self {
        Self {
            status,
            headers: HeaderMap::new(),
            body: HandlerBody::Full(Bytes::from(body)),
        }
    }
}

/// 框架无关的请求处理器（决策点 6）：鉴权 + 路由 + 单 profile/group 故障转移。
/// 默认超时值在 CLI 层解析（ProxyServerConfig.timeout 总是具体值），这里不做 Option 处理。
pub struct ProxyHandler {
    timeout: Duration,
    instance_id: String,
    group_name: Option<String>,
    profile_name: Option<String>,
    config_path: Option<std::path::PathBuf>,
    breaker: CircuitBreaker,
    forwarder: Forwarder,
    logger: ProxyLogger,
    started: Instant,
    status: Arc<RwLock<ProxyStatus>>,
}

impl ProxyHandler {
    pub fn new(config: &ProxyServerConfig) -> Self {
        Self {
            timeout: config.timeout,
            instance_id: config.instance_id.clone(),
            group_name: config.group_name.clone(),
            profile_name: config.profile_name.clone(),
            config_path: config.config_path.clone(),
            breaker: CircuitBreaker::new(),
            forwarder: Forwarder::new(),
            logger: ProxyLogger::new(&config.instance_id),
            started: Instant::now(),
            status: Arc::new(RwLock::new(ProxyStatus {
                instance_id: config.instance_id.clone(),
                kind: config.kind,
                running: true,
                host: config.host.clone(),
                port: config.port,
                group_name: config.group_name.clone(),
                active_group: config.group_name.clone(),
                profile_name: config.profile_name.clone(),
                pid: Some(std::process::id()),
                start_time: Some(swixter_core::types::now_iso()),
                ..Default::default()
            })),
        }
    }

    pub fn status(&self) -> Arc<RwLock<ProxyStatus>> {
        self.status.clone()
    }

    fn load_config(&self) -> ConfigManager {
        match &self.config_path {
            Some(p) => ConfigManager::load_from(p.clone()),
            None => ConfigManager::load(),
        }
    }

    /// TS: ProxyHandler.handleRequest（鉴权 + 路由 + access 日志 + 计数）
    pub async fn handle(
        &self,
        method: &str,
        path_and_query: &str,
        headers: &HeaderMap,
        body: &Bytes,
    ) -> HandlerResponse {
        let start = Instant::now();
        let path = path_and_query.split('?').next().unwrap_or(path_and_query);

        let resp = self
            .handle_inner(method, path, path_and_query, headers, body)
            .await;
        let ms = start.elapsed().as_millis() as u64;
        self.logger.request(method, path, resp.status, ms);
        {
            let mut s = self.status.write().unwrap();
            s.request_count += 1;
            if resp.status >= 500 {
                s.error_count += 1;
            }
        }
        let _ = crate::events::event_bus().send(crate::events::ProxyEvent::StatusUpdate(
            self.status.read().unwrap().clone(),
        ));
        resp
    }

    async fn handle_inner(
        &self,
        method: &str,
        path: &str,
        path_and_query: &str,
        headers: &HeaderMap,
        body: &Bytes,
    ) -> HandlerResponse {
        // 鉴权：除 /health 外必须 Bearer swixter-local-proxy
        if path != "/health" {
            let ok = headers
                .get("authorization")
                .and_then(|v| v.to_str().ok())
                .and_then(|v| v.strip_prefix("Bearer "))
                == Some(SWIXTER_PROXY_AUTH_TOKEN);
            if !ok {
                return HandlerResponse::json(
                    401,
                    json!({"error": "Invalid or missing proxy authentication"}),
                );
            }
        }

        if method == "GET" && path == "/health" {
            return HandlerResponse::json(
                200,
                json!({
                    "status": "ok",
                    "instanceId": self.instance_id,
                    "groupName": self.group_name,
                    "timestamp": swixter_core::types::now_iso(),
                    "uptime": self.started.elapsed().as_secs_f64(),
                }),
            );
        }

        // 路由（事实表 §端点 注册顺序；/anthropic/* 任意方法）；未匹配 404 纯文本
        let is_api_route = matches!(
            (method, path),
            ("POST", "/v1/chat/completions") | ("POST", "/v1/messages") | ("POST", "/v1/responses")
        ) || path.starts_with("/anthropic/");
        if !is_api_route {
            return HandlerResponse::text(404, "Not Found");
        }

        // 单 profile / group 分发（TS forwardToProvider；死参数 format 不保留）
        if let Some(profile_name) = &self.profile_name {
            return self
                .forward_single_profile(method, path_and_query, headers, body, profile_name)
                .await;
        }
        self.forward_group(method, path_and_query, headers, body)
            .await
    }

    /// TS: getGroup(idOrName) —— 先按 id 再按 name
    fn find_group<'a>(&self, mgr: &'a ConfigManager, id_or_name: &str) -> Option<&'a Group> {
        mgr.config()
            .groups
            .get(id_or_name)
            .or_else(|| mgr.config().groups.values().find(|g| g.name == id_or_name))
    }

    async fn forward_single_profile(
        &self,
        method: &str,
        endpoint: &str,
        headers: &HeaderMap,
        body: &Bytes,
        profile_name: &str,
    ) -> HandlerResponse {
        let mgr = self.load_config();
        let Some(profile) = mgr.get_profile(profile_name).cloned() else {
            self.logger.warn(
                "Profile not found",
                Some(json!({"profileName": profile_name})),
            );
            return HandlerResponse::json(
                503,
                json!({"error": format!("Profile not found: {profile_name}")}),
            );
        };
        match self
            .try_profile(method, endpoint, headers, body, &profile)
            .await
        {
            TryOutcome::Success(resp) => resp,
            TryOutcome::Upstream(resp) => resp, // 非 2xx 原样返回，无转移
            TryOutcome::Error(e) => {
                self.logger.error(
                    "Provider request failed",
                    Some(&e),
                    Some(json!({"profileName": profile_name})),
                );
                HandlerResponse::json(502, json!({"error": e.to_string()}))
            }
        }
    }

    /// TS: forwardGroup —— groupName 指定或 activeGroup；body 读一次复用（调用方传入 &Bytes）
    async fn forward_group(
        &self,
        method: &str,
        endpoint: &str,
        headers: &HeaderMap,
        body: &Bytes,
    ) -> HandlerResponse {
        let mgr = self.load_config();
        let group = match &self.group_name {
            Some(name) => self.find_group(&mgr, name),
            None => mgr
                .config()
                .active_group
                .as_ref()
                .and_then(|id| mgr.config().groups.get(id)),
        };
        let Some(group) = group.filter(|g| !g.profiles.is_empty()).cloned() else {
            self.logger.warn("No active group or profiles", None);
            return HandlerResponse::json(503, json!({"error": "No active group or profiles"}));
        };

        let mut errors: Vec<String> = Vec::new();
        let mut last_failure: Option<HandlerResponse> = None;

        for profile_id in &group.profiles {
            // ① 熔断 open 跳过
            if !self.breaker.is_available(profile_id) {
                self.logger.info(
                    "Skipping unavailable provider",
                    Some(json!({"profileId": profile_id})),
                );
                continue;
            }
            // ② profile 不存在跳过
            let Some(profile) = mgr.get_profile(profile_id).cloned() else {
                self.logger
                    .warn("Profile not found", Some(json!({"profileId": profile_id})));
                continue;
            };
            // ③ 格式不同且无注册转换器跳过
            let client_format = transform::infer_client_format(endpoint);
            let preset = swixter_core::presets::find_provider(&profile.provider_id);
            let target_format = transform::infer_target_api_format(&profile, preset.as_ref());
            if client_format != target_format
                && !transform::has_transformer(client_format, target_format)
            {
                self.logger.info(
                    "Skipping provider: no transformer for format pair",
                    Some(json!({"profileId": profile_id})),
                );
                continue;
            }
            // ④-⑥ transform + model 改写 + 转发
            match self
                .try_profile(method, endpoint, headers, body, &profile)
                .await
            {
                TryOutcome::Success(resp) => {
                    self.breaker.record_success(profile_id);
                    return resp;
                }
                TryOutcome::Upstream(resp) => {
                    // 5xx/429 计入熔断；其余非 2xx 只转移
                    if resp.status >= 500 || resp.status == 429 {
                        self.breaker.record_failure(profile_id);
                    }
                    errors.push(format!("{profile_id}: upstream returned {}", resp.status));
                    self.logger.warn(
                        "Provider returned upstream status",
                        Some(
                            json!({"profileId": profile_id, "status": resp.status, "fallback": true}),
                        ),
                    );
                    last_failure = Some(resp);
                }
                TryOutcome::Error(e) => {
                    self.breaker.record_failure(profile_id);
                    errors.push(format!("{profile_id}: {e}"));
                    self.logger.error(
                        "Provider request failed",
                        Some(&e),
                        Some(json!({"profileId": profile_id})),
                    );
                }
            }
        }

        self.logger.error(
            "All providers failed",
            None,
            Some(json!({"errors": errors})),
        );
        match last_failure {
            Some(resp) => resp, // 回传最后一个上游失败响应
            None => HandlerResponse::json(
                503,
                json!({"error": "All providers failed", "details": errors}),
            ),
        }
    }

    /// 单次 profile 尝试：④ transform 请求（失败回退透传原 body+原 endpoint）
    /// → ⑤ model 改写 → ⑥ 转发 → 2xx 时响应 transform（失败回退原始 body）
    async fn try_profile(
        &self,
        method: &str,
        endpoint: &str,
        headers: &HeaderMap,
        body: &Bytes,
        profile: &Profile,
    ) -> TryOutcome {
        let preset = swixter_core::presets::find_provider(&profile.provider_id);
        let client_format = transform::infer_client_format(endpoint);
        let target_format = transform::infer_target_api_format(profile, preset.as_ref());

        let mut target_endpoint = endpoint.to_string();
        let mut eff_body = body.clone();
        let mut ctx: Option<transform::TransformCtx> = None;

        if client_format != target_format {
            let parsed: Value = if body.is_empty() {
                json!({})
            } else {
                match serde_json::from_slice(body) {
                    Ok(v) => v,
                    Err(_) => json!({}),
                }
            };
            let c = transform::TransformCtx {
                endpoint: endpoint.to_string(),
                client_format,
                target_format,
                stream: parsed.get("stream").and_then(Value::as_bool) == Some(true),
            };
            match transform::transform_request(&parsed, &c) {
                Ok(t) => {
                    eff_body = Bytes::from(serde_json::to_vec(&t.body).unwrap());
                    target_endpoint = t.target_endpoint;
                    ctx = Some(c);
                }
                Err(e) => {
                    // transform 失败回退透传原 body + 原 endpoint（事实表 §Group 故障转移 ④）
                    self.logger.error(
                        "Request transform failed, falling back to passthrough",
                        Some(&e),
                        None,
                    );
                }
            }
        }

        // ⑤ model 改写
        let eff_body = crate::model::rewrite_request_body_for_profile(&eff_body, profile);

        let fwd = ForwardRequest {
            method: method.to_string(),
            path: target_endpoint,
            headers: headers.clone(),
            body: eff_body,
        };
        let resp = match self
            .forwarder
            .forward(fwd, profile, preset.as_ref(), self.timeout, target_format)
            .await
        {
            Ok(r) => r,
            Err(e) => return TryOutcome::Error(e),
        };
        if !(200..300).contains(&resp.status) {
            return TryOutcome::Upstream(into_handler_response(resp));
        }

        // 2xx：需要时响应 transform（失败回退原始 body）
        match (ctx, resp.is_stream) {
            (Some(c), true) => {
                let ForwardBody::Stream(stream) = resp.body else {
                    unreachable!()
                };
                let transformed = transform::transform_stream(stream, &c);
                TryOutcome::Success(HandlerResponse {
                    status: resp.status,
                    headers: resp.headers,
                    body: HandlerBody::Stream(transformed),
                })
            }
            (Some(c), false) => {
                let ForwardBody::Full(bytes) = resp.body else {
                    unreachable!()
                };
                let parsed: Value = if bytes.is_empty() {
                    json!({})
                } else {
                    match serde_json::from_slice(&bytes) {
                        Ok(v) => v,
                        Err(_) => json!({}),
                    }
                };
                match transform::transform_response(&parsed, &c) {
                    Ok(v) => TryOutcome::Success(HandlerResponse {
                        status: resp.status,
                        headers: resp.headers,
                        body: HandlerBody::Full(Bytes::from(serde_json::to_vec(&v).unwrap())),
                    }),
                    Err(e) => {
                        self.logger.error(
                            "Response transform failed, returning raw response",
                            Some(&e),
                            None,
                        );
                        TryOutcome::Success(HandlerResponse {
                            status: resp.status,
                            headers: resp.headers,
                            body: HandlerBody::Full(bytes),
                        })
                    }
                }
            }
            (None, true) => {
                let ForwardBody::Stream(stream) = resp.body else {
                    unreachable!()
                };
                let mapped = futures::StreamExt::map(stream, |r| r.map_err(std::io::Error::other));
                TryOutcome::Success(HandlerResponse {
                    status: resp.status,
                    headers: resp.headers,
                    body: HandlerBody::Stream(Box::pin(mapped)),
                })
            }
            (None, false) => TryOutcome::Success(into_handler_response(resp)),
        }
    }
}

enum TryOutcome {
    Success(HandlerResponse),
    Upstream(HandlerResponse), // 非 2xx 上游响应
    Error(crate::ProxyError),  // 网络异常
}

fn into_handler_response(resp: crate::forwarder::ForwardResponse) -> HandlerResponse {
    match resp.body {
        ForwardBody::Full(b) => HandlerResponse {
            status: resp.status,
            headers: resp.headers,
            body: HandlerBody::Full(b),
        },
        ForwardBody::Stream(s) => {
            let mapped = futures::StreamExt::map(s, |r| r.map_err(std::io::Error::other));
            HandlerResponse {
                status: resp.status,
                headers: resp.headers,
                body: HandlerBody::Stream(Box::pin(mapped)),
            }
        }
    }
}
